stateproof/builder.go

```diff
@@ -0,0 +1,474 @@
+	builderForRound, err := spw.makeBuilderForRound(rnd)
+	if err != nil {
+		spw.log.Warnf("addSigsToBuilder: makeBuilderForRound(%d): %v", rnd, err)
+		return
+	}
+	spw.builders[rnd] = builderForRound
+
+	for _, sig := range sigs {
+		pos, ok := builderForRound.voters.AddrToPos[sig.signer]
+		if !ok {
+			spw.log.Warnf("addSigsToBuilder: cannot find %v in round %d", sig.signer, rnd)
+			continue
+		}
+
+		isPresent, err := builderForRound.Present(pos)
+		if err != nil {
+			spw.log.Warnf("addSigsToBuilder: failed to invoke builderForRound.Present on pos %d - %w ", pos, err)
+			continue
+		}
+		if isPresent {
+			spw.log.Warnf("addSigsToBuilder: cannot add %v in round %d: position %d already added", sig.signer, rnd, pos)
+			continue
+		}
+
+		if err := builderForRound.IsValid(pos, &sig.sig, false); err != nil {
+			spw.log.Warnf("addSigsToBuilder: cannot add %v in round %d: %v", sig.signer, rnd, err)
+			continue
+		}
+		if err := builderForRound.Add(pos, sig.sig); err != nil {
+			spw.log.Warnf("addSigsToBuilder: error while adding sig. inner error: %w", err)
+			continue
+		}
+	}
+}
+
+func (spw *Worker) handleSigMessage(msg network.IncomingMessage) network.OutgoingMessage {
+	var ssig sigFromAddr
+	err := protocol.Decode(msg.Data, &ssig)
+	if err != nil {
+		spw.log.Warnf("spw.handleSigMessage(): decode: %v", err)
+		return network.OutgoingMessage{Action: network.Disconnect}
+	}
+
+	fwd, err := spw.handleSig(ssig, msg.Sender)
+	if err != nil {
+		spw.log.Warnf("spw.handleSigMessage(): %v", err)
+	}
+
+	return network.OutgoingMessage{Action: fwd}
+}
+
+// handleSig adds a signature to the pending in-memory state proof provers (builders). This function is
+// also responsible for making sure that the signature is valid, and not duplicated.
+// if a signature passes all verification it is written into the database.
+func (spw *Worker) handleSig(sfa sigFromAddr, sender network.Peer) (network.ForwardingPolicy, error) {
+	spw.mu.Lock()
+	defer spw.mu.Unlock()
+
+	builderForRound, ok := spw.builders[sfa.Round]
+	if !ok {
+		latest := spw.ledger.Latest()
+		latestHdr, err := spw.ledger.BlockHdr(latest)
+		if err != nil {
+			// The latest block in the ledger should never disappear, so this should never happen
+			return network.Disconnect, err
+		}
+
+		if sfa.Round < latestHdr.StateProofTracking[protocol.StateProofBasic].StateProofNextRound {
+			// Already have a complete state proof in ledger.
+			// Ignore this sig.
+			return network.Ignore, nil
+		}
+
+		// The sig should be for a round which is a multiple of StateProofInterval
+		// using the latestHdr protocol, since changing StateProofInterval is not supported
+		proto := config.Consensus[latestHdr.CurrentProtocol]
+
+		// proto.StateProofInterval is not expected to be 0 after passing StateProofNextRound
+		// checking anyway, otherwise will panic
+		if proto.StateProofInterval == 0 {
+			return network.Disconnect, fmt.Errorf("handleSig: StateProofInterval is 0 for round %d", latest)
```

Comment:
Similarly, this error has nothing to do with the message that was sent, I don't think that having a lot of disconnections on the network is the right action

Reply:
```suggestion
			return network.Ignore, fmt.Errorf("handleSig: StateProofInterval is 0 for round %d", latest)
```

Reply:
I disagree. In this case, a signature message was sent while stateproofs are not enabled. This user should be disconnected

Reply:
oh OK! makes sense